File: level2/전력망을 둘로 나누기.py
def solution(n, wires):
    answer = 100

    tree_root_1 = [-1 for _ in range(n+1)]

    queue = [1] * n
    check = set()
    start = 0
    end = 1

    while start < end:
        node = queue[start]
        start += 1
        cnt = 0
        for ele in wires:
            n1, n2 = ele

            if n1 in check or n2 in check:
                continue

            if n1 == node:
                queue[end] = n2
                end += 1
                cnt += 1
                continue

            if n2 == node:
                queue[end] = n1
                end += 1
                cnt += 1
                continue

        check.add(node)
        tree_root_1[node] = cnt
        for parent in queue[:start-1]:

            if not tree_root_1[parent]:
                continue

            tree_root_1[parent] += cnt

    for child in tree_root_1[1:]:
        diff = abs(n - 2*(child+1))
        if answer > diff:
            answer = diff

    print(tree_root_1)

    return answer


print(solution(9, [[1, 3], [2, 3], [3, 4], [4, 5], [4, 6], [4, 7], [7, 8], [7, 9]]))
print(solution(9, [[1, 3], [4, 5], [8, 7], [9, 7], [7, 4], [4, 6], [3, 2], [4, 3]]))
print(solution(3, [[1, 2], [3, 2]]))
print(solution(8, [[1, 2], [1, 3], [1, 4], [4, 5], [5, 6], [6, 7], [6, 8]]))

File: level2/전력망을 둘로 나누기.js
function countNode(node, subTree) {
  const queue = [node];
  const visited = new Set();
  let idx = 0;
  let length = 1;

  visited.add(node);

  while (idx < length) {
    const current = queue[idx];
    idx += 1;

    for (let nextNode = 0; nextNode < subTree.length; nextNode++) {
      const value = subTree[current][nextNode];

      if (value === 0) continue;

      if (visited.has(nextNode)) continue;

      queue.push(nextNode);
      visited.add(nextNode);
      length += 1;
    }
  }

  return queue.length;
}

function solution(n, wires) {
  let answer = 100;
  const tree = new Array(n + 1).fill(0).map(() => new Array(n + 1).fill(0))

  for (let wire of wires) {
    const [n1, n2] = wire;
    tree[n1][n2] = 1;
    tree[n2][n1] = 1;
  }

  for (let wire of wires) {
    const [n1, n2] = wire;
    tree[n1][n2] = 0;
    tree[n2][n1] = 0;

    const n1Count = countNode(n1, tree);
    const n2Count = countNode(n2, tree);
    const diff = Math.abs(n1Count - n2Count);

    if (diff < answer) answer = diff;

    tree[n1][n2] = 1;
    tree[n2][n1] = 1;
  }

  return answer;
}

console.log(solution(9, [[1, 3], [2, 3], [3, 4], [4, 5], [4, 6], [4, 7], [7, 8], [7, 9]]))
console.log(solution(2, [[1, 2]]))
console.log(solution(3, [[1, 2], [2, 3]]))